### PR TITLE
Feature/2534 update apple platform versions

### LINKED ISSUE #2534: Update Apple platform versions
Several newer Apple platform versions are not detected (correctly):

* Mac OS X versions 11 and 12
* Newer iOS versions
* iPad OS versions

Please add those to browscap.

### PR DESCRIPTION
**Pull request by paulrutter**
#2534 Update Apple platform versions
- Added 'OSX_dynamic' and 'OSX_10_16' for Mac OS
- Added 'ipadOS_15_0' / 'ipadOS_15_2'
- Added 'iOS_A_15_0' / 'iOS_A_15_2'
- Added platform to the browsers that it runs on, also including the "generic" platforms to have better coverage for future versions (when the useragent remains the same)

@asgrim @mimmi20 
I did an attempt to add the recent Apple platform versions to browscap.
Also added a "dynamic" OSX item in platforms.json

I still don't exactly understand why those "dynamic" platforms are not added to all browsers, since that would help recognizing future versions which are not specified exactly in "platforms.json". I did that now for the Apple related platforms, but not sure if that's right. Can you help me on this logic? Maybe i'm misunderstanding and those generic ones shouldn't be added?

### CODE REVIEW

**Comment by paulrutter** [comment]
Also, can you help determine which unit test if failing? 
I cannot see that in the logs: https://github.com/browscap/browscap/runs/4187437561?check_suite_focus=true

Could it be the "dynamic" platforms that i added to the browser files?

File: resources/user-agents/browsers/chrome/chrome-ios-73-on.json [hunk 1]
```diff
@@ -24,11 +24,11 @@
             "iPod": "iPod Touch"
           },
           "platforms": [
-            "iOS_A_14_2", "iOS_A_14_1", "iOS_A_14_0",
+            "iOS_A_15_2", "iOS_A_15_1", "iOS_A_15_0", "iOS_A_14_2", "iOS_A_14_1", "iOS_A_14_0",
```
Comment (mimmi20) [inline resources/user-agents/browsers/chrome/chrome-ios-73-on.json]: Please add the new versions (15.x) on a new line

File: resources/user-agents/browsers/chrome/chrome-ios-73-on.json [hunk 2]
```diff
@@ -37,11 +37,11 @@
             "iPad": "iPad"
           },
           "platforms": [
-            "ipadOS_14_2", "ipadOS_14_1", "ipadOS_14_0",
+            "ipadOS_15_2", "ipadOS_15_1", "ipadOS_15_0", "ipadOS_14_2", "ipadOS_14_1", "ipadOS_14_0",
```
Comment (mimmi20) [inline resources/user-agents/browsers/chrome/chrome-ios-73-on.json]: Please add the new versions (15.x) on a new line

File: resources/user-agents/browsers/mobile-safari/mobile-safari-generic.json [hunk 3]
```diff
@@ -22,7 +22,7 @@
             "iPod": "iPod Touch"
           },
           "platforms": [
-            "iOS_A_14_2", "iOS_A_14_1", "iOS_A_14_0",
+            "iOS_A_15_2", "iOS_A_15_1", "iOS_A_15_0", "iOS_A_14_2", "iOS_A_14_1", "iOS_A_14_0",
```
Comment (mimmi20) [inline resources/user-agents/browsers/mobile-safari/mobile-safari-generic.json]: Please add the new versions (15.x) on a new line

**Review by mimmi20 — CHANGES_REQUESTED** [review]

**Comment by paulrutter** [comment]
I fixed the comments, but i think that won't help solve the unit test issue.
Can you take a look at my question here? https://github.com/browscap/browscap/pull/2535#issuecomment-966912527 and https://github.com/browscap/browscap/pull/2535#issue-779165451

**Review by mimmi20 — APPROVED** [review]

**Comment by paulrutter** [comment]
Thanks for approving @mimmi20, but since the unit tests still fail i don't think we should merge.
I need more info on the "dynamic" platforms, as mentioned earlier.

**Comment by mimmi20** [comment]
"dynamic" platforms only work, if the platform version matches the browser version.

**Comment by mimmi20** [comment]
The Build is failing since May. Idont think your changes caused this.

**Comment by paulrutter** [comment]
> The Build is failing since May. Idont think your changes caused this.

I will try to fix the unit tests first, so we can actually see what my changes did to existing functionality.

File: resources/user-agents/browsers/chrome/chrome-generic.json [hunk 4]
```diff
@@ -48,7 +48,9 @@
           "match": "Mozilla/5.0 (#PLATFORM#)*applewebkit*(*khtml*like*gecko*)*Chrome/*",
           "device": "Macintosh",
           "platforms": [
-            "OSX_10_15", "OSX_10_14", "OSX_10_13", "OSX_10_12",
+            "OSX_12_0",
+            "OSX_11_0", "OSX_11_1",
```
Comment (mimmi20) [inline resources/user-agents/browsers/chrome/chrome-generic.json]: Please change the ordering. Please write `"OSX_11_1", "OSX_11_0"` (like you did in the Chrome file). Please change it in other files too.

File: resources/user-agents/browsers/mobile-safari/mobile-safari-15-0.json [hunk 5]
```diff
@@ -0,0 +1,66 @@
+{
+  "division": "Mobile Safari #MAJORVER#.#MINORVER#",
+  "versions": [14],
```
Comment (mimmi20) [inline resources/user-agents/browsers/mobile-safari/mobile-safari-15-0.json]: Please change the version to 15

File: resources/user-agents/browsers/mobile-safari/mobile-safari-14-0.json [hunk 6]
```diff
@@ -1,6 +1,6 @@
 {
   "division": "Mobile Safari #MAJORVER#.#MINORVER#",
-  "versions": [14],
+  "versions": ["14.0", "14.1", "14.2", "14.3", "14.4", "14.5", "14.6", 14],
```
Comment (mimmi20) [inline resources/user-agents/browsers/mobile-safari/mobile-safari-14-0.json]: Please change the order of versions. Please write the highest first. The version "14.0" is not needed, as it the same as version "14". Also change the file name to `mobile-safari-14-x.json`.

File: resources/user-agents/browsers/mobile-safari/mobile-safari-13-0.json [hunk 7]
```diff
@@ -1,6 +1,6 @@
 {
   "division": "Mobile Safari #MAJORVER#.#MINORVER#",
-  "versions": [13],
+  "versions": ["13.1", 13],
```
Comment (mimmi20) [inline resources/user-agents/browsers/mobile-safari/mobile-safari-13-0.json:3]: Please change the filename to `mobile-safari-13-x.json`.

File: resources/user-agents/browsers/safari/safari-15-0.json [hunk 8]
```diff
@@ -0,0 +1,33 @@
+{
+    "division": "Safari #MAJORVER#.#MINORVER#",
+    "versions": ["15.0", 15],
```
Comment (mimmi20) [inline resources/user-agents/browsers/safari/safari-15-0.json]: These two versions are the same, as the Zero will be added automatically, if a `#MINORVER#` placeholder is there.

File: tests/BrowscapTest/Generator/Helper/BuildHelperTest.php [hunk 9]
```diff
@@ -784,13 +784,13 @@ public function testDuplicateUseragents(): void
         $writerCollection->expects(static::once())
             ->method('renderAllDivisionsHeader')
             ->willReturnSelf();
-        $writerCollection->expects(static::exactly(2))
+            $writerCollection->expects(static::any())
```
Comment (mimmi20) [inline tests/BrowscapTest/Generator/Helper/BuildHelperTest.php]: Please keep the indent

**Review by mimmi20 — CHANGES_REQUESTED** [review]

**Comment by paulrutter** [comment]
@mimmi20 I managed to get the unit tests running again, by reverting a change in https://github.com/browscap/browscap/pull/2535/files#diff-cebc14b7503943d2278cbaf4fe86c60f57707b619511f54041399054709dd28d.

Now i had to fix few other unit tests that were added already, but didn't succeed.
I've tried, to my best knowledge, to add the new Apple platforms:

* Mac OS Big sur (11) and Monterey (12), 10.x range is no longer used although 10.16 was used at some point
* iOS 14.0 / 14.8 and 15.0 / 15.2 (which is in beta)
* iPad OS 15.0 / 15.2, didn't add 14.3 / 14.8 as i wasn't sure it was used

I also added Safari 15 specifically, as that is shipped with MacOS Monterey. 
In the readme, i updated the badge for the CI, which was no longer ok.

@mimmi20 i processed your latest comments. Can you re-review and merge if you think it's ok now?
The code coverage report still seems a bit off, but not sure how to resolve that.

**Review by mimmi20 — APPROVED** [review]

**Comment by paulrutter** [comment]
@mimmi20 can you merge and release a new browscap version if you're ok with the changes?

**Comment by mimmi20** [comment]
@paulrutter I may be able to merge, but not to release.
@asgrim Could you merge and release?

**Comment by asgrim** [comment]
Yep, thank you @paulrutter - appreciate the effort and fixing up those tests :metal:

**Comment by paulrutter** [comment]
@asgrim do we need to trigger something for http://browscap.org/ as well?

**Comment by asgrim** [comment]
Yes, waiting for Dependabot

@mimmi20 FYI the release procedure is at https://github.com/browscap/browscap/wiki/Public-release-procedure :) I believe you should be able to do most of it now

**Comment by asgrim** [comment]
@paulrutter looks like it is up now :+1:

**Comment by paulrutter** [comment]
Yes, great!

**Comment by willyaranda** [comment]
`Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_6) AppleWebKit/605.1.15 (KHTML, like Gecko) Version/15.1 Safari/605.1.15`

is incorrectly detected, as does not match. Changing `Version/15.1` for `Version/15.0` does match and get the major version properly

**Comment by paulrutter** [comment]
Hmm i wasn't aware of safari 15.1 was already released. We might add that as well.

**Comment by paulrutter** [comment]
Created https://github.com/browscap/browscap/pull/2536 @willyaranda

**Comment by AnuarT79** [comment]
Obvious

File: resources/user-agents/browsers/chrome/chrome-generic.json (hunk 4 above)
Comment (AnuarT79) [reply resources/user-agents/browsers/chrome/chrome-generic.json]: > Please change the ordering. Please write `"OSX_11_1", "OSX_11_0"` (like you did in the Chrome

**Pull request merged**